### ISSUE TITLE
Empty save() does not works as query

### ISSUE THREAD

**Issue by dreadatour** [iterative/datachain#327]
Empty `save()` in datachain query fails with `Internal error on creating dataset` error.

Query example:

```python
from datachain import DataChain

DataChain.from_storage("gs://dvcx-datacomp-small/metadata", anon=True).save()
```

This happens because if `name` is not set, [temp dataset name will be used](https://github.com/iterative/datachain/blob/main/src/datachain/query/dataset.py#L1659-L1660). At the same time, after `save()` is finishes, [temp datasets are cleaned up](https://github.com/iterative/datachain/blob/main/src/datachain/query/dataset.py#L1700).

Some explanation:

> Empty save is needed because operations are lazy. Chains are not executed immediately and intermediate results are not stored.
For example, if we want to create `dc_filtered_1` and `dc_embeddings` from `dc`, without `save()` intermediate `dc` chain will be executed twice, for each children.

Couple options here:
1. Skip saved temp dataset on cleanup
2. Use `exec()` instead of empty `save()`. Need to check if this is proper replace + need to add `exec()` to docs.

**Comment by skshetry** [comment]
Note that this happens as a part of session cleanup, not during `DatasetQuery.cleanup()`.

https://github.com/iterative/datachain/blob/0eb79599a0d774d89b5ff5a34db2b3dcba9a5016/src/datachain/query/session.py#L70

**Comment by skshetry** [comment]
If I remember correctly, this was intended to work as it is.

Btw, can you share what the "Internal Error" is? It's working fine for me on CLI.

**Comment by dreadatour** [comment]
> If I remember correctly, this was intended to work as it is.

Interesting, @mnrozhkov says it works before, also he have workflows based in this behavior. May be he can provide more details on this? 🙏

> Btw, can you share what the "Internal Error" is? It's working fine for me on CLI.

This is literally the error 👀 From this line: https://github.com/iterative/datachain/blob/0eb79599a0d774d89b5ff5a34db2b3dcba9a5016/src/datachain/catalog/catalog.py#L100

**Comment by skshetry** [comment]
Looks like this is raised on Studio side.

https://github.com/iterative/studio/blob/6e7ed3fe9b30ef2631b7cee97e5eb621263e155b/backend/datachain_worker/tasks/run_job.py#L367

**Comment by skshetry** [comment]
I think we should disable Session cleanup on Studio inside the scripts.

We can decide if we want to delete the datasets or hide it, but that should happen after the script runs.

**Comment by shcheklein** [comment]
> from datachain import DataChain
> DataChain.from_storage("gs://dvcx-datacomp-small/metadata", anon=True).save()

folks, do you remember what was the expected product behavior in this case? may be there was a discussion on this?

I feel it's better for this to be consistent with CLI? I assume in CLI we would drop the dataset (it won't be saved). Should it be the same here?

What is happening on the Studio side when we just return a datachain w/o `save()` - how do we name those tables and prevent then from being collected?

**Comment by mattseddon** [comment]
@shcheklein the expected behaviour is in the docstring:

https://github.com/iterative/datachain/blob/49141a5b069c7d13d90a1798659faacf494e1e97/src/datachain/lib/dc.py#L562-L571

**Comment by shcheklein** [comment]
okay, so, on the Studio side - what do we do with a query result (if there is no `save()` at the end) - how / where do we save it?

**Comment by skshetry** [comment]
> okay, so, on the Studio side - what do we do with a query result (if there is no `save()` at the end) - how / where do we save it?

On Studio, we always save the dataset (with `catalog.query(save=True)`.

https://github.com/iterative/datachain/blob/49141a5b069c7d13d90a1798659faacf494e1e97/src/datachain/catalog/catalog.py#L1988-L1990

We need this for showing preview results.

Here the fix is to disable cleanup on session, which fixes the error.

We can discuss what we should do about dangling datasets. But that is already the case.

**Comment by shcheklein** [comment]
thanks @skshetry! let me step back a bit:

- is this issue (in the ticket description( happening only on Studio (just to make sure I understand it right)?
- if it happening on the Studio side and we have `save=True` - what is the difference between:

`chain.save()`

and just

`chain`

why can't we just save the result in both cases in the same way and cleanup all the temp tables as we usually do?

---------

> Here the fix is to disable cleanup on session, which fixes the error.

yep, it just seems a bit strange that we have to disable GC globally to solve (what seems to me at least - I might be wrong) a "local" / specific situation with the `save()` at the end

> We need this for showing preview results.

I think btw that preview is going into a separate model / table (in addition to the result). Are we saving the resulting dataset itself besides that preview?

> But that is already the case.

could you clarify please?

**Comment by dreadatour** [comment]
Random idea: may be we should store query result dataset in session and skip it while deleting temporary datasets on session cleanup? Because it is not temporary anymore, as it is a query result. This should solve the issue.

Other possible solution might be to remove query result dataset from session temporary datasets list if it is exists in this list.

**Comment by ilongin** [comment]
This is expected behavior, just error message is bad / wrong.

Temp datasets should be used inside a query for optimization purposed, and not as a result of a query. We should just catch this error and write better error message.

**Comment by dmpetrov** [comment]
> We should just catch this error and write better error message.

Yes, it's just a messaging and docs issue. 

A radical solution - rename `save()`  without params to `persist()`

**Comment by shcheklein** [comment]
I think @ilongin @dmpetrov you are on the right track and one way is to improve the message  - probably the simplest solution here. 

Alternative - I think we can allow running a query with a "`save()` without params" at the end. It can still "persist" that dataset (or we can optimize it), return datachain and then process the result in the same way as any other query result. Not clear why it is not possible or why it is not working this way now to be honest.

We can put a warning "`save()` at the end of query is not meaningful" or something.

**Comment by dmpetrov** [comment]
> why it is not possible or why it is not working this way now to be honest.

I'm not sure about a use case. If dataset is needed - user has to assign a name.
Without params it serves a whole different purpose - avoid re-computing. That's why renaming might be best option.

**Comment by shcheklein** [comment]
> I'm not sure about the use case. If dataset is needed - user has to assign a name.
> Without params it serves a whole different purpose - avoid re-computing. That's why renaming might be best option.

Yep, I think I understand the difference. And there is not use case for this (you don't need indeed that extra "check point"). It still doesn't explain tbh why it is not working. It might be inefficient (extra table that will GCed almost right away), but it's not clear why it should not be possible.

So, if we make it permissive (allow `save()` at the end and ignore it or have that extra table):

- less chances people will be getting an exception at the very end (and have to rerun the query)
- to my mind it is more general - cleaner semantics. If `save()` returns a DC there is no immediate reason to not process it as a regular result.

If we make it strict:

- better for education - people will know that `save()` has a specific purpose (?)
- anything else??

**Comment by dmpetrov** [comment]
> It might be inefficient (extra table that will GCed almost right away), but it's not clear why it should not be possible.

I mean... yes, adjusting an error message is just a shortcut. The assumption is - we don't want spend time on this now (does not seem a priority for now) and we can assume that users do not do save() when compute-persist is not needed.

`from_storage(...).save()` or `save("mycats").save()` ideally should work just fine (a copy of a table is not even needed - can be optimized).

**Comment by shcheklein** [comment]
> from_storage(...).save() or save("mycats").save() ideally should work just fine (a copy of a table is not even needed - can be optimized).

yes, exactly

> I mean... yes, adjusting an error message is just a shortcut. 

yep, agreed if it saves time ... I would also not spend time optimizing it now or changing it if it requires any significant complicated logic.

I was thinking that it could be a small fix tbh - not sure on this.

**Comment by dmpetrov** [comment]
> I was thinking that it could be a small fix tbh - not sure on this.

If it's small fix - yes, that's wold be great. Otherwise, I'd just improve the message.

**Comment by skshetry** [comment]
Note that we also have "Register Dataset" from preview. So we do need to save the datasets in Studio (if it has to work like a detached query).

**Comment by skshetry** [comment]
Even if we do ignore that functionality, if we want to do this inside query script itself,
we'll have to add the functionality of following function inside the script:

https://github.com/iterative/datachain/blob/f0db01d7b4a9ae26fa6de1389a740d98c6ac52bc/src/datachain/catalog/catalog.py#L1144

This function saves stats, preview, schema and other information of the dataset. Note that the preview has to be raw database results, so that sys columns are also included. (CLI and Studio are currently not aligned on how they use preview results, so we'll need to also fix that).

**Comment by shcheklein** [comment]
Decided to postpone this and come back later.

**Comment by dreadatour** [comment]
Smallest fix possible for empty `save()` last line:

```diff
diff --git a/src/datachain/query/dataset.py b/src/datachain/query/dataset.py
index 027a713..7a0a79c 100644
--- a/src/datachain/query/dataset.py
+++ b/src/datachain/query/dataset.py
@@ -1778,6 +1778,8 @@ def query_wrapper(dataset_query: DatasetQuery) -> DatasetQuery:
     save = bool(os.getenv("DATACHAIN_QUERY_SAVE"))
     save_as = os.getenv("DATACHAIN_QUERY_SAVE_AS")
 
+    is_session_dataset = dataset_query.name.startswith(Session.DATASET_PREFIX)
+
     if save_as:
         if dataset_query.attached:
             dataset_name = dataset_query.name
@@ -1804,7 +1806,7 @@ def query_wrapper(dataset_query: DatasetQuery) -> DatasetQuery:
             )
         else:
             dataset_query = dataset_query.save(save_as)
-    elif save and not dataset_query.attached:
+    elif save and (is_session_dataset or not dataset_query.attached):
         name = catalog.generate_query_dataset_name()
         dataset_query = dataset_query.save(name)
 ```

**Comment by skshetry** [comment]
Closed by #357.

**Comment by dreadatour** [comment]
Follow-up issues:
- https://github.com/iterative/datachain/issues/361
- https://github.com/iterative/studio/issues/10548
